File: src-tauri/notification-bridge.js
/**
 * Notification Bridge for Tauri Mobile
 * 
 * This script intercepts web Notification API calls from the remote frontend
 * and forwards them to the native Tauri backend for display as native push notifications.
 * 
 * The script is injected into the webview and overrides the Notification constructor
 * to intercept all notification creation attempts.
 */

(function() {
    'use strict';
    
    // Check if Tauri is available
    if (typeof window.__TAURI_INTERNALS__ === 'undefined') {
        console.warn('[Tauri Notification Bridge] Tauri not available, using standard web notifications');
        return;
    }
    
    // Store original Notification constructor
    const OriginalNotification = window.Notification;
    
    // Get Tauri invoke function
    let invoke;
    try {
        const tauri = window.__TAURI__;
        if (tauri && tauri.tauri && tauri.tauri.invoke) {
            invoke = tauri.tauri.invoke.bind(tauri.tauri);
        } else {
            console.warn('[Tauri Notification Bridge] Tauri invoke not available');
            return;
        }
    } catch (e) {
        console.warn('[Tauri Notification Bridge] Failed to get Tauri API:', e);
        return;
    }
    
    // Override Notification constructor
    const TauriNotification = function(title, options = {}) {
        const notification = {
            title: String(title),
            body: options.body || '',
            icon: options.icon || null,
            tag: options.tag || null,
            data: options.data || null,
            requireInteraction: options.requireInteraction || false,
            silent: options.silent || false,
            timestamp: Date.now(),
        };
        
        // Show native notification via Tauri
        showNativeNotification(notification).catch(err => {
            console.error('[Tauri Notification Bridge] Failed to show native notification:', err);
            // Fallback to original notification if available
            if (OriginalNotification && OriginalNotification.permission === 'granted') {
                try {
                    new OriginalNotification(notification.title, {
                        body: notification.body,
                        icon: notification.icon
                    });
                } catch (e) {
                    console.error('[Tauri Notification Bridge] Fallback notification also failed:', e);
                }
            }
        });
        
        // Create a mock notification object for compatibility
        const mockNotification = {
            title: notification.title,
            body: notification.body,
            icon: notification.icon,
            tag: notification.tag,
            data: notification.data,
            onclick: null,
            onshow: null,
            onerror: null,
            onclose: null,
            close: function() {
                // Native notifications can't be programmatically closed
                if (this.onclose) {
                    this.onclose();
                }
            },
            addEventListener: function(event, handler) {
                if (event === 'click') {
                    this.onclick = handler;
                } else if (event === 'show') {
                    this.onshow = handler;
                } else if (event === 'error') {
                    this.onerror = handler;
                } else if (event === 'close') {
                    this.onclose = handler;
                }
            },
            removeEventListener: function(event, handler) {
                // Remove event handlers if needed
            }
        };
        
        // Trigger onshow if provided
        if (options.onshow) {
            setTimeout(() => {
                try {
                    options.onshow(mockNotification);
                } catch (e) {
                    console.error('[Tauri Notification Bridge] Error in onshow handler:', e);
                }
            }, 0);
        }
        
        return mockNotification;
    };
    
    // Copy static properties
    Object.defineProperty(TauriNotification, 'permission', {
        get: function() {
            // Check permission status (will be updated by requestPermission)
            return window.__TAURI_NOTIFICATION_PERMISSION__ || 'default';
        },
        set: function(value) {
            window.__TAURI_NOTIFICATION_PERMISSION__ = value;
        },
        configurable: true
    });
    
    Object.defineProperty(TauriNotification, 'maxActions', {
        get: function() {
            return OriginalNotification ? OriginalNotification.maxActions : 0;
        },
        configurable: true
    });
    
    // Implement requestPermission
    TauriNotification.requestPermission = function(callback) {
        const promise = invoke('request_notification_permission')
            .then(granted => {
                const permission = granted ? 'granted' : 'denied';
                window.__TAURI_NOTIFICATION_PERMISSION__ = permission;
                return permission;
            })
            .catch(err => {
                console.error('[Tauri Notification Bridge] Permission request failed:', err);
                const permission = 'denied';
                window.__TAURI_NOTIFICATION_PERMISSION__ = permission;
                return permission;
            });
        
        if (typeof callback === 'function') {
            promise.then(callback);
            return undefined;
        }
        
        return promise;
    };
    
    // Function to show native notification via Tauri
    async function showNativeNotification(notification) {
        try {
            await invoke('show_notification', {
                title: notification.title,
                body: notification.body,
                icon: notification.icon
            });
            
            console.log('[Tauri Notification Bridge] Native notification shown:', notification.title);
        } catch (error) {
            console.error('[Tauri Notification Bridge] Failed to show native notification:', error);
            throw error;
        }
    }
    
    // Replace global Notification
    window.Notification = TauriNotification;
    
    // Initialize permission status
    invoke('is_notification_supported')
        .then(supported => {
            if (supported) {
                // Request permission on initialization
                TauriNotification.requestPermission().then(permission => {
                    console.log('[Tauri Notification Bridge] Notification permission:', permission);
                });
            }
        })
        .catch(err => {
            console.warn('[Tauri Notification Bridge] Failed to check notification support:', err);
        });
    
    console.log('[Tauri Notification Bridge] Initialized successfully');
})();

File: src-tauri/src/notification_bridge.rs
/// Notification bridge module
///
/// This module provides functionality to convert web notifications
/// from the remote frontend into native push notifications.

use tauri::AppHandle;
use crate::notifications;

/// Show a native notification
///
/// This command receives notification data from the frontend and displays
/// it as a native notification on iOS/Android using platform-specific APIs.
///
/// # Arguments
///
/// * `app` - The Tauri app handle
/// * `title` - Notification title
/// * `body` - Notification body text
/// * `icon` - Optional icon URL or path (used on Android)
///
/// # Returns
///
/// Returns `Ok(())` on success, or an error if the operation fails.
#[tauri::command]
pub async fn show_notification(
    _app: AppHandle,
    title: String,
    body: String,
    icon: Option<String>,
) -> Result<(), String> {
    log::info!("Showing native notification: {} - {}", title, body);
    
    // Use platform-specific notification implementation
    notifications::show_notification(
        &title,
        &body,
        icon.as_deref(),
    )
}

/// Request notification permissions
///
/// On mobile platforms, notification permissions are requested from the system.
/// This command requests permission and returns the result.
///
/// # Returns
///
/// Returns `true` if permission is granted, `false` otherwise.
#[tauri::command]
pub async fn request_notification_permission(_app: AppHandle) -> Result<bool, String> {
    log::info!("Requesting notification permission");
    
    // Use platform-specific permission request
    notifications::request_permission()
}

/// Check if notifications are supported
///
/// # Returns
///
/// Returns `true` if notifications are supported on this platform.
#[tauri::command]
pub async fn is_notification_supported() -> Result<bool, String> {
    // Notifications are supported on both iOS and Android
    #[cfg(any(target_os = "ios", target_os = "android"))]
    {
        Ok(true)
    }
    
    #[cfg(not(any(target_os = "ios", target_os = "android")))]
    {
        Ok(false)
    }
}

/// Check notification permission status
///
/// # Returns
///
/// Returns `true` if permission is granted, `false` otherwise.
#[tauri::command]
pub async fn check_notification_permission(_app: AppHandle) -> Result<bool, String> {
    log::info!("Checking notification permission status");
    
    // Use platform-specific permission check
    notifications::check_permission()
}

